src/include/slab_containers.h

```diff
@@ -0,0 +1,728 @@
+// if the container doesn't exceed "stackSize" nodes.
+//
+// Subsequent slabs are allocated using the supplied allocator. A slab on the heap, by default, contains 'slabSize' nodes.
+// However, a "reserve" function (same functionality as vector::reserve) is provided that ensure a minimum number
+// of free nodes is available without further memory allocation. If the slab_xxxx::reserve function needs to allocate
+// additional nodes, only a single memory allocation will be done. Meaning that it's possible to have slabs that
+// are larger (or smaller) than 'slabSize' nodes.
+//
+template<pool_index_t pool_ix,typename T,size_t stackSize, size_t slabSize>
+class slab_allocator : public pool_slab_allocator<pool_ix,T> {
+  struct slab_t;
+  struct slabHead_t {
+    slabHead_t *prev;
+    slabHead_t *next;
+  };
+  //
+  // Each slot has a pointer to it's containing slab PLUS either one Object OR if it's free exactly one pointer 
+  // in a per-slab freelist.
+  // Since this is raw memory, we don't want to declare something of type "T" to avoid
+  // accidental constructor/destructor calls.
+  //
+  struct slot_t {
+    slab_t *slab; // Pointer to my slab, NULL for slots on the stack (Technically, this could be an index but I'm lazy :))
+    enum { OBJECT_IN_POINTERS = (sizeof(T) + sizeof(void *) - 1) / sizeof(void *) };
+    slot_t *storage[OBJECT_IN_POINTERS]; // Either one object OR index[0] points to next element in free list for this slab
+  };
+   //
+   // Each slab has a freelist of objects within the slab and the size of that list
+   // The size is needed to cheaply determine when all of the slots within a slab are unused
+   // so that we can free the slab itself    
+   //
+   struct slab_t {
+      slabHead_t slabHead;  // membership of slab in list of slabs with free elements
+      uint32_t size;    // # of allocated slots in this slab
+      uint32_t freeSlots;   // # of free slots, i.e., size of freelist of slots within this slab
+      slot_t *freeHead;     // Head of list of freeslots OR NULL if none
+      slot_t slot[0];       // slots
+   };
+   slab_t *slabHeadToSlab(slabHead_t *head) { return reinterpret_cast<slab_t *>(head); }
+   //
+   // Exactly one of these as part of this container (the slab on the stack :))
+   //
+   struct stackSlab_t : public slab_t {
+      slot_t stackSlot[stackSize]; // Allows the compiler to get the right size :)
+   };
+   //
+   // Initialize a new slab, remember, "T" might not be correct use the stored sizes.
+   //
+   void initSlab(slab_t *slab,size_t sz) {
+      slab->size = sz;
+      slab->freeSlots = 0; // Pretend that it was completely allocated before :)
+      slab->freeHead = NULL;
+      slab->slabHead.next = NULL;
+      slab->slabHead.prev = NULL;
+      this->slab_allocate(sz,trueSlotSize,sizeof(slab_t));
+      char *raw = reinterpret_cast<char *>(slab->slot);
+      for (size_t i = 0; i < sz; ++i) {
```

Comment:
Bringing up a slab with even moderately high (32) number of slots is costly. Move from freelist to bitmap for `slot_t` tracking? For the sake of simplicity having 64 as an upper limit appears acceptable.

```
       │           s->storage[0] = slab->freeHead;                             ▒
       │           slab->freeHead = s;                                         ▒
 41,45 │       mov    %rdx,0x80(%rbp) 
```
where self is 2,94% cycles of `msgr-worker-0`:
```
-    3,60%     2,94%  msgr-worker-0  ceph-osd  [.] WeightedPriorityQueue<OpQueueItem, unsigned long>::Queue::insert                                                                                          ▒
```

Reply:
After switching to `bitmap` and lazy `slot_t::slab` assignment:

```
-    1,80%     0,80%  msgr-worker-0  ceph-osd  [.] WeightedPriorityQueue<OpQueueItem, unsigne▒
   - 1,00% WeightedPriorityQueue<OpQueueItem, unsigned long>::Queue::insert                  ◆
      - 0,58% operator new                                                                   ▒
         - tcmalloc::ThreadCache::FetchFromCentralCache                                      ▒
            + tcmalloc::CentralFreeList::RemoveRange                                         ▒
        0,24% __popcountdi2                                                                  ▒
        0,05% boost::intrusive::rbtree_algorithms<boost::intrusive::rbtree_node_traits<void*,▒
        0,03% __popcountdi2@plt                                                              ▒
        0,02% std::__detail::_List_node_base::_M_hook                                        ▒
```

Still, running with the atomic updates just commented-out.